arch/rv32i/src/lib.rs

```diff
@@ -192,16 +186,85 @@ extern "C" {
+    ///
+    /// In the case of an interrupt while the kernel was executing we only need
+    /// to save the kernel registers and then run whatever interrupt handling
+    /// code we need to. If the trap happens while and application was
+    /// executing, we have to save the application state and then resume the
+    /// `switch_to()` function to correctly return back to the kernel.
+    ///
+    /// We implement this distinction through a branch on the value of the
+    /// `mscratch` CSR value. If, at the time the trap was taken, it contains
+    /// `0`, we assume that the target is currently executing kernel code.
+    ///
+    /// If it contains any other value, we interpret it to be a memory address
+    /// pointing to a particular structure:
+    ///
+    /// ```
+    /// mscratch           0               1               2               3
+    ///  \->|--------------------------------------------------------------|
+    ///     | scratch word, overwritten with s1 register contents          |
+    ///     |--------------------------------------------------------------|
+    ///     | trap handler address, continue trap handler execution here   |
+    ///     |--------------------------------------------------------------|
+    /// ```
+    ///
+    /// Thus, process implementations can define their own strategy for how
+    /// traps should be handled when they occur during process execution.
+    /// This global trap handler behavior is well defined. It will:
+    ///
+    /// 1. atomically swap s0 and the mscratch CSR,
+    ///
+    /// 2. execute the default kernel trap handler if s0 now contains `0`
+    /// (meaning that the mscratch CSR contained `0` before entering this trap
+    /// handler),
+    ///
+    /// 3. otherwise, save s1 to `0*4(s0)`, and finally
+    ///
+    /// 4. load the address at `1*4(s0)` into s1, and jump to it.
+    ///
+    /// No registers other than s0, s1 and the mscratch CSR are to be clobbered
+    /// before contiuing execution at the address loaded into the mscratch CSR
+    /// or the _start_kernel_trap kernel trap handler.  Execution with these
+    /// second-stage trap handlers must continue in the same trap handler
+    /// context as originally invoked by the trap (e.g., the global trap handler
+    /// will not execute an mret instruction). It will not modify CSRs that
+    /// contain information on the trap source or the system state prior to
+    /// entering the trap handler.
+    ///
+    /// We explicitly use callee-saved over temporary registers, as this makes
+    /// it easier to call other functions as part of the trap, handler, such as
+    /// to disable interrupts. This global trap handler saves the previous
+    /// values of these registers anyways, and places new, important values into
+    /// them (the previous mscratch value and the trap handler address). These
+    /// values are likely desired to not be clobbered by further function calls
+    /// in the trap handler.
```

Comment:
```suggestion
    /// We deliberately clobber callee-saved instead of caller-saved registers, as this makes
    /// it easier to call other functions as part of the trap handler (for example to
    /// to disable interrupts from within Rust code). This global trap handler saves the previous
    /// values of these clobbered registers ensuring that they can be restored later.
    /// It places new values into these clobbered registers (such as the previous `s0` register contents)
    /// that are required to be retained for correctly returning from the trap handler,
    /// and as such need to be saved across C-ABI function calls. Loading them into
    /// saved registers avoids the need to manually save them across such calls.
```